sklearn/calibration.py

```diff
@@ -658,3 +660,232 @@ def calibration_curve(y_true, y_prob, *, normalize=False, n_bins=5,
+    to create a visualizer. All parameters are stored as attributes.
+
+    Read more in the :ref:`User Guide <calibration>`.
+
+    Parameters
+    -----------
+    prob_true : ndarray
+        The proportion of samples whose class is the positive class (fraction
+        of positives), in each bin.
+
+    prob_pred : ndarray
+        The mean predicted probability in each bin.
+
+    y_prob : ndarray of shape (n_samples,)
+        Probability estimates for the positive class, for each sample.
+
+    brier_value : int or None
+        The Brier score value. If `None`, the Brier score is not shown.
+
+    estimator_name : str, default=None
+        Name of estimator. If `None`, then the estimator name is not shown.
+
+    Attributes
+    ----------
+    line_ : matplotlib Artist
+        Calibration curve.
+
+    ax_ : matplotlib Axes
+        Axes with calibration curve.
+
+    figure_ : matplotlib Figure
+        Figure containing the curve.
+    """
+    def __init__(self, prob_true, prob_pred, y_prob, *,
+                 brier_value=None, estimator_name=None):
+        self.prob_true = prob_true
+        self.prob_pred = prob_pred
+        self.y_prob = y_prob
+        self.brier_value = brier_value
+        self.estimator_name = estimator_name
+
+    def plot(self, ax=None, *, name=None, ref_line=True, **kwargs):
+        """Plot visualization.
+
+        Extra keyword arguments will be passed to matplotlib's `plot`.
+
+        Parameters
+        ----------
+        ax : Matplotlib Axes, default=None
+            Axes object to plot on. If `None`, a new figure and axes is
+            created.
+
+        name : str, default=None
+            Name of calibration curve for labeling. If `None`, use the
+            name of the estimator.
+
+        ref_line : bool, default=True
+            If `True`, plots a reference line representing a perfectly
+            calibrated classifier.
+
+        **kwargs : dict
+            Keyword arguments to be passed to :func:`matplotlib.pyplot.plot`.
+
+        Returns
+        -------
+        display : :class:`~sklearn.metrics.CalibrationDisplay`
+            Object that stores computed values.
+        """
+        check_matplotlib_support("CalibrationDisplay.plot")
+        import matplotlib.pyplot as plt
+
+        if ax is None:
+            fig, ax = plt.subplots()
+
+        name = self.estimator_name if name is None else name
+
+        line_kwargs = {}
+        if self.brier_value is not None and name is not None:
+            line_kwargs["label"] = \
+                f"{name} (Brier: {self.brier_value:0.3f})"
+        elif self.brier_value is not None:
+            line_kwargs["label"] = f"Brier: {self.brier_value:0.3f}"
+        elif name is not None:
+            line_kwargs["label"] = name
+        line_kwargs.update(**kwargs)
+
+        existing_ref_line = ('Perfectly calibrated' in
+                             ax.get_legend_handles_labels()[1])
```

Comment:
This is quite sneaky and clever.

Reply:
> Which would mean the default could be False.

I can change the default to False.

> Or a more robust solution would be to store a ref_line_ attribute (could be private) which would be None if there's no ref line yet

Yes that is more robust, I will change to this

Reply:
> Which would mean the default could be False.

Hmm actually the parameter is designed such that it is only drawn once on a plot, so even if it is True, if there is an existing reference line, it is not drawn again. So the default can stay True.

Reply:
Yes we can leave the default to True if we're still manually preventing the line to plotted twice. I was just saying that if we remove this mechanism (which I think is fair), it would be best to chance the default to False.

Feel free to do whichever you prefer. If we leave it True, I'd rather rely on an attribute rather than on the label, which can be changed.

Reply:
Private with `_` or `__` ?

Reply:
just one leading underscore is enough (no trailing underscore)

Reply:
I have just realised we can't do this with an attribute. When draw a second line we only pass the `ax_` attribute of the `CalibrationDisplay`.

E.g., form `test_plot_calibration_curve_ref_line`:
```Python
def test_plot_calibration_curve_ref_line(pyplot, data_binary):
    # Check that `ref_line` only appears once
    X, y = data_binary
    lr = LogisticRegression().fit(X, y)
    dt = DecisionTreeClassifier().fit(X, y)

    viz = plot_calibration_curve(lr, X, y)
    viz2 = plot_calibration_curve(dt, X, y, ax=viz.ax_)

    labels = viz2.ax_.get_legend_handles_labels()[1]
    assert labels.count('Perfectly calibrated') == 1
```

The current method would fail (only) if the user makes a plot, renames the label then plots another line. In this case they could set `ref_line=False` and draw their own line.

Is there another way to do this that I am missing? We can scrap `ref_line` as well.

Reply:
@thomasjpfan isn't there a Visualizer where we store private attributes in an Axes or Figure instance?

Reply:
ping @thomasjpfan

Reply:
```
$ git grep "self.ax_"
doc/developers/plotting.rst:           self.ax_ = ax
sklearn/metrics/_plot/confusion_matrix.py:        self.ax_ = ax
sklearn/metrics/_plot/precision_recall_curve.py:        self.ax_ = ax
sklearn/metrics/_plot/roc_curve.py:        self.ax_ = ax
```

Reply:
I was more referring something like `self.ax_._has_ref_line` ;)

Reply:
Indeed, I misread your comment...

Reply:
@NicolasHug and @thomasjpfan to follow up on this, I had a look at the other Visualizer classes and I didn't see a case of a private attribute being stored in Axes or Figure instance.
Also, to clarify, this would/could be done with:
```
setattr(ax, '_ref_line', True)
```

I think the current implementation is reasonably robust. It would fail only when a user re-names the reference line and then plots another curve e.g., plot curve, re-name reference line, plot another curve.
But I am happy to implement the private attribute, please let me know!

Reply:
OK, let's keep it this way then.

Reply:
I have a PR that was used to store a weak reference to the visualizer: https://github.com/scikit-learn/scikit-learn/pull/15702